gpsdecode.1
.\"     Title: gpsdecode
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 03/18/2009
.\"    Manual: 13 Jul 2005
.\"    Source: 13 Jul 2005
.\"
.TH "GPSDECODE" "1" "03/18/2009" "13 Jul 2005" "13 Jul 2005"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
gpsdecode \- decode RTCM or AIVDM streams into a readable format
.SH "SYNOPSIS"
.HP 10
\fBgpsdecode\fR [\-d] [\-e] [\-l] [\-u] [\-D\ \fIdebuglevel\fR] [\-V]
.SH "DESCRIPTION"
.PP
This tool is a decoder/encoder for various binary packet formats associated with GPS and differential\-correction services\&. It produces a text dump on standard output from binary on standard input, or binary packets on standard output from text on standard input, and aims to be 100% information\-preserving in both directions\&. As well as data, the decoder also prints decoder status messages to standard error as necessary\&.
.PP
Two of the supported formats are RTCM 2 and 3, a pair of obscure and complicated serial protocol used for broadcasting pseudorange corrections from differential\-GPS reference stations\&. You can use this mode of the tool with
\fBnc\fR(1)
to examine RTCM feeds from DGPSIP servers or Ntrip broadcasters\&. The decoder dump format for RTCM2 is described in
\fBrtcm\fR(5); these lines go to standard output\&.
.PP
Another supported format is AIVDM\&. This is the sentence format used by the marine Automatic Identification System\&. This can be decoded, but not yet encoded\&.
.PP
The dump lines are values of AIS payload fields, comma\-separated, in the order that they occur in the payload, spare and assigned\-mode bits omitted\&. Spans of fields expressing a date are emitted as an ISO8601 timestamp (look for colons and the trailing Z indicating Zulu/UTC time), and the 19\-bit group of TDMA status fields found at the end of message types 1\-4 are are dumped as a single unsigned integer in hex\&. (Unused regional\-authority fields are also dumped in hex\&.)
.PP
By default, certain scaling and conversion operations are performed for the output\&. Latitudes and longitudes are scaled to decimal degrees rather than the native AIS unit of 1/10000th of a minute of arc\&. Ship (but not air) speeds are scaled to knots rather than tenth\-of\-knot unit\&. Navigation status and positioning\-system type are dumped as text strings rather than IAS numeric codes\&. Rate of turn may appear as "nan" if is unavailable, of as one of the strings "fastright" or "fastleft" if it is out of the IAS encoding range; otherwise it is quadratically mapped back to the turn sensor number in degrees per minute\&. Vessel draughts are converted to decimal meters rather than native AIS decimeters\&.
.SH "OPTIONS"
.PP
The
\fB\-d\fR
option tells the program to decode packets presented on standard input to a text dump on standard output\&. This is the default behavior\&.
.PP
RTCM2 will be dumped in the format of
\fBrtcm-104\fR(5)
on standard output\&.
.PP
The
\fB\-e\fR
option option tells the program to encode a text dump in the format of
\fBrtcm-104\fR(5)
to standard output\&. Encoding is supported only RTCM2, not for RTCM3\&.
.PP
The
\fB\-u\fR
suppresses scaling to float quantities and text expansion of numeric codes\&. A dump with this option is lossless\&.
.PP
The
\fB\-l\fR
inserts a mnemonic label, followed by an equal sign, before each field in each dump line\&.
.PP
The
\fB\-V\fR
option directs the program to emit its version number, then exit\&.
.PP
The
\fB\-D\fR
option sets a debug verbosity level\&. It is mainly of interest to developers\&.
.SH "APPLICABLE STANDARDS"
.PP
The applicable standard for V2 is
RTCM Recommended Standards for Differential NAVSTAR GPS Service
RTCM Paper 194\-93/SC 104\-STD\&.
.PP
Note that
gpsdecode
presently recognizes only the 2\&.1 level of RTCM; the protocol was revised up to a version 2\&.3 including additional messages relating to GLONASS and real\-time kinematics before being deprecated in favor of V3\&. It is now obsolete\&.
.PP
The applicable standard for V3 is
RTCM Standard 10403\&.1 for Differential GNSS Services \- Version 3
RTCM Paper 177\-2006\-SC104\-STD\&.
.PP
Ordering instructions for the RTCM standards are accessible from the website of the
\fIRadio Technical Commission for Maritime Services\fR\&[1]
under "Publications"\&.
.PP
The applicable standard for AIVDM is
ITU\-R M\&.1371: ITU Recommendation on the Technical Characteristics for a Universal Shipborne Automatic Identification System (AIS) using Time Division Multiple Access in the Maritime Mobile Band,\&. A more accessible description can be found onn the references page of the GPSD project website\&.
.SH "BUGS AND LIMITATIONS"
.PP
AIDVM decoding presently only parses fields for the common sentence types 1\-5, the rarely encountered type 9, and the uncommon type 18\&. AIVDM encoding is not yet supported\&.
.PP
RTCM3 decoding is buggy and incomplete\&.
.PP
RTCM2 represents floating\-point quantities as an integer multiple of a fixed scale factor\&. Editing an RTCM2 dump can produce numbers that are not an integer multiple of the scale factor for their field\&. If you do this, the value actually packed into binary RTCM2 will be rounded down to the nearest scale unit, and dumping will show slightly different numbers than those you entered\&. This bug could be fixed by supporting the
\fB\-u\fR
option to suppress scaling, but the protocol is noew obsolete\&.
.PP
The RTCM2 decoder logic is sufficiently convoluted to confuse some compiler optimizers, notably in GCC 3\&.x at \-O2, into generating bad code\&.
.SH "SEE ALSO"
.PP

\fBgpsd\fR(8),
\fBgps\fR(1),
\fBlibgps\fR(3),
\fBlibgpsd\fR(3),
\fBgpsprof\fR(1),
\fBgpsfake\fR(1),
\fBrtcm-104\fR(5)\&.
.SH "AUTHOR"
.PP
Eric S\&. Raymond
<esr@thyrsus\&.com>\&. This is a somewhat hacked version of an RTCM decoder originally written by Wolfgang Rupprecht\&. There is a project page for
gpsd
\fIhere\fR\&[2]\&.
.SH "NOTES"
.IP " 1." 4
Radio Technical Commission for Maritime Services
.RS 4
\%http://www.rtcm.org/
.RE
.IP " 2." 4
here
.RS 4
\%http://gpsd.berlios.de/
.RE


debian/changelog
gpsd (2.90.1~svn6819-1) unstable; urgency=low

  * SVN snapshot, uploading to unstable as it fixes
    the blocking issues with gps_poll.
    The non-blocking gps_poll was introduced while fixing
    an issue in gps_unpack, so this upload also
    Closes: #561280
  * [84ab918f] Merge branch 'master' into experimental.
  * [bbd1d183] Merge branch 'experimental-upstream' into experimental
  * [fd98f498] Update symbols file.
  * [6f63fa61] Update gbp.conf.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 16 Dec 2009 10:25:26 +0100

gpsd (2.90-2) unstable; urgency=low

  * Upload to unstable.
    This will also fix the following issues:
    - gpsd doesn't support pipelining (Closes: #488913)
    - FTBFS with binutils-gold (Closes: #554739) 

 -- Bernd Zeimetz <bzed@debian.org>  Tue, 08 Dec 2009 13:51:27 +0100

gpsd (2.90-1) experimental; urgency=low

  * New upstream release.
  * [bb9c3eb5] Add Conflicts: fso-gpsd to libgps19. fso-gpsd is a mess
    and not supported.
  * [4b2d3b57] Merge commit 'upstream/2.90'
  * [fdad4622] Work around --disable-python not being available anymore.

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 07 Dec 2009 11:13:14 +0100

gpsd (2.90~svn6694-1) experimental; urgency=low

  * SVN snapshot.
  * [10b2460a] Include gps.h in Tachometer.c, needed to define UNUSED.
  * [3f75872e] Merge branch 'experimental-upstream' into experimental
  * [33831f96] Update gbp config for experimental.
  * [476e547e] Update soname in debian/control.
  * [0561c6cf] Updaing patches, dropping those which were applied
    upstream.
  * [d721c055] Revert "Revert "Remove workarounds for broken --disable-
    python."" This reverts commit
    0e5c78b75ec0390b9317c4e8db3f9294ae9cf319. Conflicts:  debian/rules
  * [151e52e5] Clean up patches which were applied by upstream.
  * [b9bff239] Merge branch 'experimental-upstream' into experimental
  * [981b2ce1] Ship gps.1 again as the package `gps' was RMed from
    Debian.
  * [9674fb87] Add a Conflicts: gps for gpsd-clients. Just in case the
    package is still installed somewhere.
  * [6c788d8d] Clean ../gpsd_config.h.in~ in case it was left by the
    build.
  * [cf72c26f] Delete packaging/gpsd.spec.in.rej, was done upstream
    also.
  * [5eadc03b] Dropping INSTALL-description-fix. Not necessary anymore.
  * [c4a1d3f1] python-gps needs python2.6 or simplejson, update Depends.
  * [5e8ed26d] Do not try to install manpages via gpsd-clients.install.
  * [2b598581] s/dj_installman/dh_installman/
  * [1fc3151e] Update symbols files for the new soname.
  * [f530af1b] Install missing manpages which were forgotten before.
  * [9a964d22] Make hotplug wrapper executable.
  * [06b11aa2] Add python-gps.manpages to install manpages.

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 03 Dec 2009 21:12:54 +0100

gpsd (2.39-8) unstable; urgency=low

  * [a9e9d167] Export needed variables for the hotplug script. (Closes:
    #556956) - thanks to Luca Niccoli
  * [254f1da4] Add a missing % which caused USB autostart to fail.
    (Closes: #556954) - thanks to Luca Niccoli
  * [2990d2cc] Updating debconf translation: pt.po (Closes: #556250) -
    thanks to "Traduz - Portuguese Translation Team"
  * [dc6305e7] Updating debconf translation: sv.po (Closes: #556580) -
    thanks to Martin Bagge
  * [6066918d] Add missing : in po files and unfuzzy them.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 18 Nov 2009 17:39:31 +0100

gpsd (2.39-7) unstable; urgency=low

  [ Bernd Zeimetz ]
  * [bda49bd9] Bumping Standards-Version to 3.8.3. No changes needed.
  * [cdf3c74c] Fix hotplug script environment patch to ensure that
    'None' is not passed as gpsd option.
  * [67cde41b] Don't check for "$START_DAEMON" = "true" in
    gpsd.hotplug.wrapper anymore.
  * [c4fb8bab] Go trough all debconf questions if the daemon is started
    at boot *or* USB autodetection is enabled.
  * [f929f33f] Add dbus to 'Suggests' of gpsd.
  * [9dd1656f] Rewriting gpsd.init completely. The new start-stop-daemon
    options should ensure that gpsd is stopped definitely, hopefully
    this fixes an upgrade problem which was reported on Launchpad. (LP:
    #468915) - thanks to Walter Wittel
  * [376285e7] Updating gpsd.templates after review. - thanks to Justin
    B Rye
  * [e5d5e1f7] Updating po files for new gpsd.templates.
  * [6e23af3e] Update po files after running debconf-updatepo.
  * [691c564c] Add missing : to the control path question.
  * [8da65d0a] Running debconf-updatepo again.

  [ Esko Arajärvi ]
  * [77a66dc2] Updating debconf translation: fi.po (Closes: #554478)

  [ Kenshi Muto ]
  * [389d8f08] Updating debconf translation: ja.po (Closes: #554576)

  [ dooteo ]
  * [400eada8] Updating debconf translation: eu.po (Closes: #554540)

  [ Ivan Masár ]
  * [67146d71] Updating debconf translation: sk.po (Closes: #554681)

  [ Luca Monducci ]
  * [c52ba7fa] Updating debconf translation: it.po (Closes: #555117)

  [ Eddy Petrișor ]
  * [afb4ef0b] Updating debconf translation: ro.po (Closes: #555464)

  [ Thomas Mueller ]
  * [9f7c21e9] Updating debconf translation: de.po (Closes: #555840)

  [ Miroslav Kure ]
  * [e4246738] Updating debconf translation: cs.po (Closes: #555894)

  [ Francisco Javier Cuadrado ]
  * [3a72944f] Updating debconf translation: es.po (Closes: #556035)

  [ Yuri Kozlov ]
  * [4beef045] Updating debconf translation: ru.po (Closes: #556193)

  [ marce ]
  * [325806a3] Updating debconf translation: gl.po (Closes: #555453)

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 14 Nov 2009 21:43:26 +0100

gpsd (2.39-6) experimental; urgency=low

  [ Christian Perrier ]
  * [fe381184] Updating debconf translation: fr.po (Closes: #540502)

  [ Jordà Polo ]
  * [2c23a9a6] Updating debconf translation: ca.po (Closes: #544101)

  [ Clytie Siddall ]
  * [49af80a6] Updating debconf translation: vi.po (Closes: #548020)

  [ Bernd Zeimetz ]
  * [35fe08a1] Correct dependencies in init.d LSB header. (Closes:
    #548552) - thanks to Petter Reinholdtsen <pere@hungry.com>

gpsdecode.c
/* $Id: gpsdecode.c 6623 2009-11-30 01:37:49Z esr $ */
#include <sys/types.h>
#ifndef S_SPLINT_S
#include <unistd.h>
#endif /* S_SPLINT_S */
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdbool.h>
#include <ctype.h>

#include "gpsd.h"
#include "gps_json.h"

static int verbose = 0;
static bool scaled = true;
static bool json = false;

/**************************************************************************
 *
 * Generic machinery
 *
 **************************************************************************/

void gpsd_report(int errlevel, const char *fmt, ... )
/* assemble command in printf(3) style, use stderr or syslog */
{
    if (errlevel <= verbose) {
	char buf[BUFSIZ];
	va_list ap;

	(void)strlcpy(buf, "gpsdecode: ", BUFSIZ);
	va_start(ap, fmt) ;
	(void)vsnprintf(buf + strlen(buf), sizeof(buf)-strlen(buf), fmt, ap);
	va_end(ap);
	(void)fputs(buf, stdout);
    }
}

static void aivdm_csv_dump(struct ais_t *ais, char *buf, size_t buflen)
{
    (void)snprintf(buf, buflen, "%u,%u,%09u,", ais->type,ais->repeat,ais->mmsi);
    /*@ -formatcode @*/
    switch (ais->type) {
    case 1:	/* Position Report */
    case 2:
    case 3:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%d,%u,%u,%d,%d,%u,%u,%u,0x%x,%u,0x%x",
		       ais->type1.status,
		       ais->type1.turn,
		       ais->type1.speed,
		       (uint)ais->type1.accuracy,
		       ais->type1.lon,
		       ais->type1.lat,
		       ais->type1.course,
		       ais->type1.heading,
		       ais->type1.second,
		       ais->type1.maneuver,
		       (uint)ais->type1.raim,
		       ais->type1.radio);
	break;
    case 4:	/* Base Station Report */
    case 11:	/* UTC/Date Response */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%04u:%02u:%02uT%02u:%02u:%02uZ,%u,%d,%d,%u,%u,0x%x",
		       ais->type4.year,
		       ais->type4.month,
		       ais->type4.day,
		       ais->type4.hour,
		       ais->type4.minute,
		       ais->type4.second,
		       (uint)ais->type4.accuracy,
		       ais->type4.lon,
		       ais->type4.lat,
		       ais->type4.epfd,
		       (uint)ais->type4.raim,
		       ais->type4.radio);
	break;
    case 5: /* Ship static and voyage related data */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%s,%s,%u,%u,%u,%u,%u,%u,%02u-%02uT%02u:%02uZ,%u,%s,%u",
		       ais->type5.imo,
		       ais->type5.ais_version,
		       ais->type5.callsign,
		       ais->type5.shipname,
		       ais->type5.shiptype,
		       ais->type5.to_bow,
		       ais->type5.to_stern,
		       ais->type5.to_port,
		       ais->type5.to_starboard,
		       ais->type5.epfd,
		       ais->type5.month,
		       ais->type5.day,
		       ais->type5.hour,
		       ais->type5.minute,
		       ais->type5.draught,
		       ais->type5.destination,
		       ais->type5.dte);
	break;
    case 6:	/* Binary Message */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u,%zd:%s",
		       ais->type6.seqno,
		       ais->type6.dest_mmsi,
		       (uint)ais->type6.retransmit,
		       ais->type6.app_id,
		       ais->type6.bitcount,
		       gpsd_hexdump(ais->type6.bitdata,
				       (ais->type6.bitcount+7)/8));
	break;
    case 7:	/* Binary Acknowledge */
    case 13:	/* Safety Related Acknowledge */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u",
		       ais->type7.mmsi1,
		       ais->type7.mmsi2,
		       ais->type7.mmsi3,
		       ais->type7.mmsi4);
	break;
    case 8:	/* Binary Broadcast Message */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%zd:%s",
		       ais->type8.app_id,
		       ais->type8.bitcount,
		       gpsd_hexdump(ais->type8.bitdata,
				       (ais->type8.bitcount+7)/8));
	break;
    case 9:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%d,%d,%u,%u,0x%x,%u,%u,0x%x",
		       ais->type9.alt,
		       ais->type9.speed,
		       (uint)ais->type9.accuracy,
		       ais->type9.lon,
		       ais->type9.lat,
		       ais->type9.course,
		       ais->type9.second,
		       ais->type9.regional,
		       ais->type9.dte,
		       (uint)ais->type9.raim,
		       ais->type9.radio);
	break;
    case 10:	/* UTC/Date Inquiry */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u",
		       ais->type10.dest_mmsi);
	break;
    case 12:	/* Safety Related Message */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%s",
		       ais->type12.seqno,
		       ais->type12.dest_mmsi,
		       (uint)ais->type12.retransmit,
		       ais->type12.text);
	break;
    case 14:	/* Safety Related Broadcast Message */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%s", 
		       ais->type14.text);
	break;
    case 15:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u,%u,%u,%u,%u",
		       ais->type15.mmsi1,
		       ais->type15.type1_1,
		       ais->type15.offset1_1,
		       ais->type15.type1_2,
		       ais->type15.offset1_2,
		       ais->type15.mmsi2,
		       ais->type15.type2_1,
		       ais->type15.offset2_1);
	break;
    case 16:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u,%u,%u",
		       ais->type16.mmsi1,
		       ais->type16.offset1,
		       ais->type16.increment1,
		       ais->type16.mmsi2,
		       ais->type16.offset2,
		       ais->type16.increment2);
	break;
    case 17:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%d,%d,%zd:%s",
		       ais->type17.lon,
		       ais->type17.lat,
		       ais->type17.bitcount,
		       gpsd_hexdump(ais->type17.bitdata,
				       (ais->type17.bitcount+7)/8));
	break;
    case 18:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%d,%d,%u,%u,%u,0x%x,%u,%u,%u,%u,%u,%u,0x%x",
		       ais->type18.reserved,
		       ais->type18.speed,
		       (uint)ais->type18.accuracy,
		       ais->type18.lon,
		       ais->type18.lat,
		       ais->type18.course,
		       ais->type18.heading,
		       ais->type18.second,
		       ais->type18.regional,
		       (uint)ais->type18.cs,
		       (uint)ais->type18.display,
		       (uint)ais->type18.dsc,
		       (uint)ais->type18.band,
		       (uint)ais->type18.msg22,
		       (uint)ais->type18.raim,
		       ais->type18.radio);
	break;
    case 19:
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%d,%d,%u,%u,%u,0x%x,%s,%u,%u,%u,%u,%u,%u,%u,%u,%u",
		       ais->type19.reserved,
		       ais->type19.speed,
		       (uint)ais->type19.accuracy,
		       ais->type19.lon,
		       ais->type19.lat,
		       ais->type19.course,
		       ais->type19.heading,
		       ais->type19.second,
		       ais->type19.regional,
		       ais->type19.shipname,
		       ais->type19.shiptype,
		       ais->type19.to_bow,
		       ais->type19.to_stern,
		       ais->type19.to_port,
		       ais->type19.to_starboard,
		       ais->type19.epfd,
		       (uint)ais->type19.raim,
		       ais->type19.dte,
		       (uint)ais->type19.assigned);
	break;
    case 20:	/* Data Link Management Message */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u,%u",
		       ais->type20.offset1,
		       ais->type20.number1,
		       ais->type20.timeout1,
		       ais->type20.increment1,
		       ais->type20.offset2,
		       ais->type20.number2,
		       ais->type20.timeout2,
		       ais->type20.increment2,
		       ais->type20.offset3,
		       ais->type20.number3,
		       ais->type20.timeout3,
		       ais->type20.increment3,
		       ais->type20.offset4,
		       ais->type20.number4,
		       ais->type20.timeout4,
		       ais->type20.increment4);
	break;
    case 21: /* Aid to Navigation */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%s,%u,%d,%d,%u,%u,%u,%u,%u,%u,%u,0x%x,%u,%u",
		       ais->type21.aid_type,
		       ais->type21.name,
		       (uint)ais->type21.accuracy,
		       ais->type21.lon,
		       ais->type21.lat,
		       ais->type21.to_bow,
		       ais->type21.to_stern,
		       ais->type21.to_port,
		       ais->type21.to_starboard,
		       ais->type21.epfd,
		       ais->type21.second,
		       ais->type21.regional,
		       (uint)ais->type21.off_position,
		       (uint)ais->type21.raim,
		       (uint)ais->type21.virtual_aid);
	break;
    case 22:	/* Channel Management */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,%u,%u,%u,%d,%d,%d,%d,%u,%u,%u,%u",
		       ais->type22.channel_a,
		       ais->type22.channel_b,
		       ais->type22.txrx,
		       (uint)ais->type22.power,
		       ais->type22.ne_lon,
		       ais->type22.ne_lat,
		       ais->type22.sw_lon,
		       ais->type22.sw_lat,
		       (uint)ais->type22.addressed,
		       (uint)ais->type22.band_a,
		       (uint)ais->type22.band_b,
		       ais->type22.zonesize);
	break;
    case 23:	/* Group Management Command*/
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%d,%d,%d,%d,%u,%u,%u,%u,%u",
		       ais->type23.ne_lon,
		       ais->type23.ne_lat,
		       ais->type23.sw_lon,
		       ais->type23.sw_lat,
		       ais->type23.stationtype,
		       ais->type23.shiptype,
		       ais->type23.txrx,
		       ais->type23.interval,
		       ais->type23.quiet);
	break;
    case 24: /* Class B CS Static Data Report */
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf), 
		       "%s,",
		       ais->type24.shipname);
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%u,",
		       ais->type24.shiptype);
	(void)snprintf(buf+strlen(buf), buflen-strlen(buf),
		       "%s,%s,",
		       ais->type24.vendorid,
		       ais->type24.callsign);
	if (AIS_AUXILIARY_MMSI(ais->mmsi)) {
	    (void)snprintf(buf+strlen(buf), buflen-strlen(buf),
			   "%u",
			   ais->type24.mothership_mmsi);
	} else {
	    (void)snprintf(buf+strlen(buf), buflen-strlen(buf),
			   "%u,%u,%u,%u",
			   ais->type24.dim.to_bow,
			   ais->type24.dim.to_stern,
			   ais->type24.dim.to_port,
			   ais->type24.dim.to_starboard);
	}
	break;
    default:
	(void)snprintf(buf+strlen(buf),
		      buflen-strlen(buf), "unknown AIVDM message content.");
	break;
    }
    /*@ +formatcode @*/
    (void)strlcat(buf, "\r\n", buflen);
}

/*@ -compdestroy -compdef -usedef @*/
static void decode(FILE *fpin, FILE *fpout)
/* RTCM or AIS packets on fpin to dump format on fpout */
{
    struct gps_packet_t lexer;
    struct rtcm2_t rtcm2;
    struct rtcm3_t rtcm3;
    struct ais_t ais;
    struct aivdm_context_t aivdm;
    char buf[BUFSIZ];

    packet_reset(&lexer);

    while (packet_get(fileno(fpin), &lexer) > 0) {
	if (lexer.type == COMMENT_PACKET)
	    continue;
	else if (lexer.type == RTCM2_PACKET) {
	    rtcm2_unpack(&rtcm2, (char *)lexer.isgps.buf);
	    if (json)
		rtcm2_json_dump(&rtcm2, buf, sizeof(buf));
	    else
		rtcm2_sager_dump(&rtcm2, buf, sizeof(buf));
	    (void)fputs(buf, fpout);
	}
	else if (lexer.type == RTCM3_PACKET) {
	    rtcm3_unpack(&rtcm3, (char *)lexer.outbuffer);
	    rtcm3_dump(&rtcm3, stdout);
	}
	else if (lexer.type == AIVDM_PACKET) {
	    if (verbose >=1 )
		(void)fputs((char *)lexer.outbuffer, stdout);
	    /*@ -uniondef */
	    if (aivdm_decode((char *)lexer.outbuffer, lexer.outbuflen, &aivdm, &ais)) {
		if (!json)
		    aivdm_csv_dump(&ais, buf, sizeof(buf));
		else
		    aivdm_json_dump(&ais, scaled, buf, sizeof(buf));
		(void)fputs(buf, fpout);
	    }
	    
	    /*@ +uniondef */
	}
    }
}
/*@ +compdestroy +compdef +usedef @*/

/*@ -compdestroy @*/
static void encode(FILE *fpin, bool repack, FILE *fpout)
/* dump format on fpin to RTCM-104 on fpout */
{
    char inbuf[BUFSIZ];
    struct gps_data_t gpsdata;
    int lineno = 0;

    memset(&gpsdata, '\0', sizeof(gpsdata));	/* avoid segfault due to garbage in thread-hook slots */
    while (fgets(inbuf, (int)sizeof(inbuf), fpin) != NULL) {
	int status;

	++lineno;
	if (inbuf[0] == '#')
	    continue;
	status = libgps_json_unpack(inbuf, &gpsdata, NULL);
	if (status != 0) {
	    (void) fprintf(stderr, "gpsdecode: bailing out with status %d (%s) on line %d\n", status, json_error_string(status), lineno);
	    exit(1);
	} if ((gpsdata.set & RTCM2_SET) != 0) { 
	    if (repack) {
		// FIXME: This code is presently broken
		struct gps_packet_t lexer;
		(void)memset(lexer.isgps.buf, 0, sizeof(lexer.isgps.buf));
	        (void)rtcm2_repack(&gpsdata.rtcm2, lexer.isgps.buf);
	        if (fwrite(lexer.isgps.buf, 
		       sizeof(isgps30bits_t), 
		       (size_t)gpsdata.rtcm2.length, fpout) != (size_t)gpsdata.rtcm2.length)
		    (void) fprintf(stderr, "gpsdecode: report write failed.\n");
		memset(&lexer, 0, sizeof(lexer));
	    } else {
		/* this works */
		char outbuf[BUFSIZ];
		rtcm2_json_dump(&gpsdata.rtcm2, outbuf, sizeof(outbuf));
		(void)fputs(outbuf, fpout);
	    }
	} if ((gpsdata.set & AIS_SET) != 0) { 
	    char outbuf[BUFSIZ];
	    aivdm_json_dump(&gpsdata.ais, false, outbuf, sizeof(outbuf));
	    (void)fputs(outbuf, fpout);
	}
    }
}
/*@ +compdestroy @*/

int main(int argc, char **argv)
{
    int c;
    enum {doencode, dodecode} mode = dodecode;

    while ((c = getopt(argc, argv, "cdejpuVD:")) != EOF) {
	switch (c) {
	case 'c':
	    json = false;
	    break;

	case 'd':
	    mode = dodecode;
	    break;

	case 'e':
	    mode = doencode;
	    break;

	case 'j':
	    json = true;
	    break;

	case 'u':
	    scaled = false;
	    break;

	case 'D':
	    verbose = atoi(optarg);
	    gpsd_hexdump_level = verbose;
	    break;

	case 'V':
	    (void)fprintf(stderr, "SVN ID: $Id: gpsdecode.c 6623 2009-11-30 01:37:49Z esr $ \n");
	    exit(0);

	case '?':
	default:
	    (void)fputs("gpsdecode [-v]\n", stderr);
	    exit(1);
	}
    }
    argc -= optind;
    argv += optind;

    if (mode == doencode)
	encode(stdin, !json, stdout);
    else
	decode(stdin, stdout);
    exit(0);
}

/* gpsdecode.c ends here */
